File: 16-Trie/cpp/palindrome_pairs.cpp
/*
 * Problem: Palindrome Pairs
 * Source:  LeetCode 336 - https://leetcode.com/problems/palindrome-pairs/
 * 
 * Approach: Use a hashmap for O(1) reverse lookups. For each word, check:
 *           1. If reverse of entire word exists (different index)
 *           2. For each split, if one part is palindrome and reverse of other exists
 * Time Complexity:  O(n * k^2) where k = avg word length
 * Space Complexity: O(n * k)
 * 
 * Tags: trie, hashing, palindrome, string
 */

#include <bits/stdc++.h>
using namespace std;

bool isPalin(const string& s, int left, int right) {
    while (left < right) {
        if (s[left++] != s[right--]) return false;
    }
    return true;
}

vector<vector<int>> palindromePairs(vector<string>& words) {
    unordered_map<string, int> wordMap;
    int n = words.size();
    for (int i = 0; i < n; i++) wordMap[words[i]] = i;
    
    vector<vector<int>> result;
    
    for (int i = 0; i < n; i++) {
        string word = words[i];
        int len = word.size();
        
        for (int j = 0; j <= len; j++) {
            // Case 1: word[0..j-1] is palindrome, check if reverse of word[j..len-1] exists
            if (isPalin(word, 0, j - 1)) {
                string suffix = word.substr(j);
                string rev(suffix.rbegin(), suffix.rend());
                if (wordMap.count(rev) && wordMap[rev] != i) {
                    result.push_back({wordMap[rev], i});
                }
            }
            
            // Case 2: word[j..len-1] is palindrome, check if reverse of word[0..j-1] exists
            if (j < len && isPalin(word, j, len - 1)) {
                string prefix = word.substr(0, j);
                string rev(prefix.rbegin(), prefix.rend());
                if (wordMap.count(rev) && wordMap[rev] != i) {
                    result.push_back({i, wordMap[rev]});
                }
            }
        }
    }
    
    return result;
}

int main() {
    vector<string> words = {"abcd", "dcba", "lls", "s", "sssll"};
    auto result = palindromePairs(words);
    
    cout << "Palindrome pairs:" << endl;
    for (auto& p : result) {
        cout << "[" << p[0] << ", " << p[1] << "] -> \"" 
             << words[p[0]] + words[p[1]] << "\"" << endl;
    }
    
    return 0;
}

File: 16-Trie/java/PalindromePairs.java
/*
 * Problem: Palindrome Pairs
 * Source:  LeetCode 336 - https://leetcode.com/problems/palindrome-pairs/
 * 
 * Approach: HashMap-based. For each word, check all prefix/suffix splits.
 * Time Complexity:  O(n * k^2)
 * Space Complexity: O(n * k)
 * 
 * Tags: trie, hashing, palindrome, string
 */

import java.util.*;

public class PalindromePairs {
    
    boolean isPalin(String s, int left, int right) {
        while (left < right) {
            if (s.charAt(left++) != s.charAt(right--)) return false;
        }
        return true;
    }
    
    public List<List<Integer>> palindromePairs(String[] words) {
        Map<String, Integer> wordMap = new HashMap<>();
        for (int i = 0; i < words.length; i++) wordMap.put(words[i], i);
        
        List<List<Integer>> result = new ArrayList<>();
        
        for (int i = 0; i < words.length; i++) {
            String word = words[i];
            int len = word.length();
            
            for (int j = 0; j <= len; j++) {
                // Case 1: prefix is palindrome, check reverse of suffix
                if (isPalin(word, 0, j - 1)) {
                    String rev = new StringBuilder(word.substring(j)).reverse().toString();
                    if (wordMap.containsKey(rev) && wordMap.get(rev) != i) {
                        result.add(Arrays.asList(wordMap.get(rev), i));
                    }
                }
                
                // Case 2: suffix is palindrome, check reverse of prefix
                if (j < len && isPalin(word, j, len - 1)) {
                    String rev = new StringBuilder(word.substring(0, j)).reverse().toString();
                    if (wordMap.containsKey(rev) && wordMap.get(rev) != i) {
                        result.add(Arrays.asList(i, wordMap.get(rev)));
                    }
                }
            }
        }
        
        return result;
    }
    
    public static void main(String[] args) {
        PalindromePairs sol = new PalindromePairs();
        String[] words = {"abcd", "dcba", "lls", "s", "sssll"};
        System.out.println("Palindrome pairs: " + sol.palindromePairs(words));
    }
}
